diff --git a/arch/arm64/boot/dts/ti/Makefile b/arch/arm64/boot/dts/ti/Makefile
index 1013e496ef06..fdeed660dde2 100644
--- a/arch/arm64/boot/dts/ti/Makefile
+++ b/arch/arm64/boot/dts/ti/Makefile
@@ -148,6 +148,8 @@ dtb-$(CONFIG_ARCH_K3) += k3-j721s2-evm-pcie1-ep.dtbo
 
 # Boards with J722s SoC
 dtb-$(CONFIG_ARCH_K3) += k3-j722s-evm.dtb
+dtb-$(CONFIG_ARCH_K3) += k3-j722s-ecu1270.dtb
+dtb-$(CONFIG_ARCH_K3) += k3-j722s-ecu1270-dio.dtb
 dtb-$(CONFIG_ARCH_K3) += k3-j722s-evm-csi2-ov5640.dtbo
 dtb-$(CONFIG_ARCH_K3) += k3-j722s-evm-csi2-quad-rpi-cam-imx219.dtbo
 dtb-$(CONFIG_ARCH_K3) += k3-j722s-evm-csi2-quad-tevi-ov5640.dtbo
diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index 625af75833fc..80e6256f8e2e 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -248,6 +248,12 @@ config SONYPI
 	  To compile this driver as a module, choose M here: the
 	  module will be called sonypi.
 
+config ADV_BOARD
+	tristate "Advantech board driver"
+	default n
+	help
+	  Advantech board driver
+
 config MWAVE
 	tristate "ACP Modem (Mwave) support"
 	depends on X86 && TTY
diff --git a/drivers/char/Makefile b/drivers/char/Makefile
index c5f532e412f1..2b87fcc65527 100644
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -44,3 +44,4 @@ obj-$(CONFIG_PS3_FLASH)		+= ps3flash.o
 obj-$(CONFIG_XILLYBUS_CLASS)	+= xillybus/
 obj-$(CONFIG_POWERNV_OP_PANEL)	+= powernv-op-panel.o
 obj-$(CONFIG_ADI)		+= adi.o
+obj-$(CONFIG_ADV_BOARD)		+= adv_board.o
diff --git a/drivers/net/phy/dp83867.c b/drivers/net/phy/dp83867.c
index e397e7d642d9..8a5f5255d0d8 100644
--- a/drivers/net/phy/dp83867.c
+++ b/drivers/net/phy/dp83867.c
@@ -815,6 +815,9 @@ static int dp83867_config_init(struct phy_device *phydev)
 			val &= ~DP83867_PHYCR_RESERVED_MASK;
 
 		ret = phy_write(phydev, MII_DP83867_PHYCTRL, val);
+		if (ret)
+			return ret;
+		ret = phy_write(phydev, DP83867_LEDCR1, 0x6B56);
 		if (ret)
 			return ret;
 
-- 
2.34.1

